src/ic_sign.rs

```diff
@@ -0,0 +1,221 @@
+use std::fmt;
+
+use candid::Principal;
+use ethers_core::k256::elliptic_curve::sec1::ToEncodedPoint;
+use ethers_core::k256::PublicKey;
+use ethers_core::types::transaction::eip2718::TypedTransaction;
+use ethers_core::types::{Signature, SignatureError, H160};
+use ethers_core::utils;
+use ic_canister::virtual_canister_call;
+use ic_exports::ic_ic00_types::{
+    DerivationPath, ECDSAPublicKeyArgs, ECDSAPublicKeyResponse, EcdsaCurve, EcdsaKeyId,
+    SignWithECDSAArgs,
+};
+use ic_exports::ic_kit::RejectionCode;
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum IcSignerError {
+    #[error("IC failed to sign data with rejection code {0:?}: {1}")]
+    SigningFailed(RejectionCode, String),
+
+    #[error("from address is not specified in transaction")]
+    FromAdderessNotPresent,
+
+    #[error("invalid public key")]
+    InvalidPublicKey,
+
+    #[error(transparent)]
+    SignatureError(#[from] SignatureError),
+}
+
+/// Signing key which will be used by management canister.
+#[derive(Debug, Clone, Copy)]
+pub enum SigningKeyId {
+    /// A default key ID that is used in deploying to a local version of IC (via DFX).
+    Dfx,
+
+    /// A master test key ID that is used on the mainnet.
+    Test,
+
+    /// A master production key ID that is used on the mainnet.
+    Production,
+}
+
+impl fmt::Display for SigningKeyId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            SigningKeyId::Dfx => write!(f, "dfx_test_key"),
+            SigningKeyId::Test => write!(f, "test_key_1"),
+            SigningKeyId::Production => write!(f, "key_1"),
```

Comment:
Is this correct?

Reply:
Yes: 
https://internetcomputer.org/docs/current/samples/t-ecdsa-sample#update-source-code-with-the-right-key-id

```
There are three options:

- dfx_test_key: a default key ID that is used in deploying to a local version of IC (via IC SDK).
- test_key_1: a master test key ID that is used on the mainnet.
- key_1: a master production key ID that is used on the mainnet.
```

Reply:
Done